### PR TITLE
aws - core - notify action dynamic payload

### LINKED ISSUE #8592: aws - notify action batch_size flexibility
### Describe the feature

Currently, notify action batch_size is set to 250. Some of our policies are hitting `SNS:MessagetooLongError` error even after lowering the batch size to 100 especially on ecs-task-definition/ecs-services policies.
Does is make sense to have this as a parameter in the policy within the notify action? That way we could stick to the batch_size of 250 in general and modify policies to have lower values as needed. The lower batch_size then would apply to a fewer set of policies instead of having a lower threshold across the board

### Extra information or context

_No response_

### PR DESCRIPTION
**Pull request by timmygrable**
Closes #8592 

@kapilt Is this the approach you were looking for in issue #8592 ?

TODO:

- [ ] Tests

### CODE REVIEW

File: c7n/actions/notify.py [hunk 1]
```diff
@@ -302,6 +295,42 @@ def send_sqs(self, message):
             MessageBody=self.pack(message),
             MessageAttributes=attrs)
         return result['MessageId']
+    
+    def process_batches(self, resources, message):
+        packed_resources = []
+        for resource in resources:
+            # Make a list of tuples containing the compressed resource and size
+            prepared_resource = self.prepare_resources([resource])
```
Comment (PratMis) [inline c7n/actions/notify.py:303]: This could potentially happen outside the loop, no?

**Comment by PratMis** [comment]
@kapilt thanks for explanation on this the other day. We discussed this internally and @timmygrable wanted to work on it 🙂. When you have few minutes please let us know your thoughts? And thanks for the PR @timmygrable

**Comment by PratMis** [comment]
@kapilt, thanks for your feedback. I have tried to address a few things you pointed out.
1. Moved the dynamic buffer generation to the base class `BaseNotify`
2. There's a [pack](https://github.com/cloud-custodian/cloud-custodian/blob/main/c7n/actions/notify.py#L39) function already present that serializes + compress the message (zlib.compress with default compression level Z_DEFAULT_COMPRESSION represents a default compromise between speed and compression currently equivalent to level 6). I have used that function to get the size of the compressed messgae.
3. I have also taken into consideration the size of the message in general (that includes some metadata including the actual policy). Some of our policies are extremely huge. 
4. Based on the default size threshold, `dynamic_resource_batch` is constructing a payload and sending it back.
5. Posting some logs re: the policy that was hitting this issue. For the sake of visibility I also logged in the size of each message 
```
2023-06-23 10:15:21,763: custodian.cache:DEBUG Disabling cache
2023-06-23 10:15:21,763: custodian.commands:DEBUG Loaded file ecs.yml. Contains 1 policies
2023-06-23 10:15:21,949: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-06-23 10:15:22,224: custodian.output:DEBUG Storing output with <LogFile file://output/ecs-service/custodian-run.log>
2023-06-23 10:15:22,331: custodian.policy:DEBUG Running policy:ecs-service resource:ecs-service region:us-east-1 c7n:0.9.27
2023-06-23 10:15:22,424: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-06-23 10:15:22,707: custodian.resources.ecscluster:DEBUG Filtered from 88 to 88 ecscluster
2023-06-23 10:15:28,137: custodian.resources.service:DEBUG Filtered from 112 to 112 service
2023-06-23 10:15:28,137: custodian.policy:INFO policy:ecs-service resource:ecs-service region:us-east-1 count:112 time:5.81
2023-06-23 10:15:29,849: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-06-23 10:15:30,326: custodian.actions:INFO sent message:15ea61d1-83a1-XXXX-XXXX-XXXXXXXXXX policy:ecs-service template:default count:56
2023-06-23 10:15:31,650: botocore.credentials:INFO Found credentials in shared credentials file: ~/.aws/credentials
2023-06-23 10:15:32,025: custodian.actions:INFO sent message:f7da9aff-95f8-XXXX-XXXX-XXXXXXXXXX policy:ecs-service template:default count:56
2023-06-23 10:15:32,026: custodian.policy:INFO policy:ecs-service action:notify resources:112 execution_time:3.81
2023-06-23 10:15:32,028: custodian.output:DEBUG metric:ResourceCount Count:112 policy:ecs-service restype:ecs-service scope:policy
2023-06-23 10:15:32,028: custodian.output:DEBUG metric:ApiCalls Count:160 policy:ecs-service restype:ecs-service
```
Please let me know what do you think wrt the above changes, Happy to address. And as usual thanks for all the help 🙂

**Review by ajkerrigan — COMMENTED** [review]
Since this is a pretty core change I'd want some other eyes on it, but looks good to me. I do think we'd want a test to exercise the batching. If you already have a live case of hitting the limit that might be useful, otherwise maybe a batch of mock resources seems reasonable.

**Comment by PratMis** [comment]
> Since this is a pretty core change I'd want some other eyes on it, but looks good to me. I do think we'd want a test to exercise the batching. If you already have a live case of hitting the limit that might be useful, otherwise maybe a batch of mock resources seems reasonable.

Thanks @ajkerrigan and agreed re: core change 😅 . We do have live cases of hitting the limit that we could leverage to push out tests 😄 .

**Comment by PratMis** [comment]
Spent some time thinking about writing tests. Something like this works, but it captures a lot of data. @kapilt whenever you get a chance, would love to get your thoughts. Thanks
```
    def test_notify_dynamic_payload(self):
        session_factory = self.replay_flight_data("test_notify_dynamic_payload")
        sqs_client = session_factory().client("sqs")
        queue_url = "https://sqs.us-east-1.amazonaws.com/644160558196/test-message-too-long"

        if self.recording:
            time.sleep(80)

        policy = self.load_policy(
            {
                "name": "s3-bucket",
                "resource": "s3",
                "actions": [
                    {
                        "type": "notify",
                        "to": ["someon@example.com"],
                        "transport": {
                            "type": "sqs",
                            "queue": queue_url,
                        },
                    }
                ],
            },
            config={"region": "us-east-1"},
            session_factory=session_factory,
        )

        resources = policy.poll()
        message_count = sqs_client.get_queue_attributes(
            QueueUrl=queue_url, AttributeNames=["ApproximateNumberOfMessages"]
        ).get("Attributes").get("ApproximateNumberOfMessages")
        message_list = []
        self.assertEqual(message_count, '2')
        m1 = sqs_client.receive_message(
            QueueUrl=queue_url, AttributeNames=["All"]
        ).get(
            "Messages", []
        )
        m2 = sqs_client.receive_message(
            QueueUrl=queue_url, AttributeNames=["All"]
        ).get(
            "Messages", []
        )
        message_list.append(m1)
        message_list.append(m2)
        total_resources = 0
        for message in message_list:
            body = json.loads(zlib.decompress(base64.b64decode(message[0]["Body"])))
            self.assertEqual(
                set(body.keys()),
                {
                    "account_id",
                    "partition",
                    "version",
                    "execution_id",
                    "execution_start",
                    "action",
                    "event",
                    "policy",
                    "region",
                    "account",
                    "resources",
                },
            )
            self.assertTrue(policy.resource_manager.actions[0].get_size(message) < 262144)
            total_resources += len(body.get('resources'))
        self.assertEqual(total_resources, len(resources))
```

**Pull request closed without merging**